### DISCUSSION TITLE
How to specify user name and password for a FastKafka connection

### DISCUSSION THREAD

**Discussion by Jrqz** [airtai/fastkafka#6] [Q&A] [answered]
Hi, I have inherited a partly/mostly written python Unix app which uses this FastKafka module. I was given a user name and password which the app's Producer presumably needs to send messages to the remote Kafka server. But nothing I can see in the FaskKafka code or documentation, nor AIOKafka which I gather FastKafka wraps, mentions user names and passwords!

I know little about Kafka, but I gather there is something called SASL which can configure this via a file. I would prefer to use environment variables if possible, and specify these items in the code like my other connect parameters, e.g.:

`    def get_default_config(self):

        kafka_brokers: dict = {

            "local": {

                "description": "local development kafka broker",

                "url": os.environ.get("KAFKA_BROKER_URL", "localhost"),

                "port": os.environ.get("KAFKA_BROKER_PORT", 9092),

                "protocol": "kafka-secure",

                "security": {"type": "plain"},

                "username": os.environ.get("KAFKA_BROKER_USERNAME", None),   # ???!

                "password": os.environ.get("KAFKA_BROKER_PASSWORD", None),   # ???!

            },

        }

        return {

            "kafka_brokers": kafka_brokers,

            "title": "SF Ingestion",

        }`

(Although this config says "local broker", the URL is actually remote, in a development staging area.)

I am not fussy though. So I would very much appreciate it is someone could indicate the recommended way to specify these items. TIA

**Comment by kumaranvpl** [comment]
Hello @Jrqz,

I have noticed that your staging Kafka setup likely includes security measures such as usernames, passwords, and encryption. In light of this, the code to initialize the FastKafka object should resemble the following:

    import os
    from fastkafka import FastKafka
    from aiokafka.helpers import create_ssl_context
    
    kafka_brokers = {
        "local": {
            "description": "local development kafka broker",
            "url": os.environ.get("KAFKA_BROKER_URL", "localhost"),
            "port": os.environ.get("KAFKA_BROKER_PORT", 9092),
            "protocol": "kafka-secure",
            "security": {"type": "scramSha256"},
        },
    }

    fastkafka_app = FastKafka(
        title="Demo Kafka app",
        security_protocol=os.environ["KAFKA_SECURITY_PROTOCOL"],
        sasl_mechanism=os.environ["KAFKA_SASL_MECHANISM"],
        sasl_plain_username=os.environ["KAFKA_BROKER_USERNAME"],
        sasl_plain_password=os.environ["KAFKA_BROKER_PASSWORD"],
        ssl_context=create_ssl_context(),
        kafka_brokers=kafka_brokers,
    )

In the provided code, the `kafka_brokers` dictionary is utilized for generating [AsyncAPI](https://www.asyncapi.com/) specifications and documentation. The actual Kafka credentials should be specified during the initialization of the FastKafka object. Additionally, it's important to ensure accurate values are provided for the `security_protocol` and `sasl_mechanism` parameters, as these can differ based on the configuration of your Kafka instances.

I'd like to note that FastKafka is built on top of the AIOKafka project. If you wish to pass more arguments to AIOKafka, you can easily refer to [AIOKafka's producer documentation](https://aiokafka.readthedocs.io/en/stable/api.html#aiokafka.AIOKafkaProducer) and supply these arguments as keyword arguments during the initialization of the FastKafka object.

Please review the code and make any necessary adjustments based on your specific setup. If you have any further questions or concerns, feel free to ask.

**Reply by Jrqz** [reply]
Very many thanks for your prompt reply Kumaran. I made all those changes, but I still see the same error from FastKafka, as follows.

I tried studying the code in producer_decorator.py, although this is fairly unfathomable! It appears that at some point a list called 'producer_store' is created, in which the "producer" variable is taken (line 221) as the second element. But a debug statement I added shows that only the first element is not None, hence the error.

```23-08-17 09:51:41.715 [INFO] sf_ingestion.service_logging: Producing case feed event to chatoptin-sf-connector topic: event_id=UUID('fb365675-35bd-4062-9a71-828d59c425f0') event_run_id=UUID('2e33da1f-12ad-4b2d-be30-d0fd29d6a880') event_creation_timestamp=datetime.datetime(2023, 8, 17, 8, 51, 41, 715325) case_id='01122348' case_creation_timestamp=datetime.datetime(2023, 7, 20, 10, 37, 11, tzinfo=TzInfo(UTC)) blob_URI=Url('https://eonchatoptinstorage.blob.core.windows.net/sf-input-raw/01122348.json')
Traceback (most recent call last):
  File "/app/sf_ingestion/main.py", line 43, in <module>
    loop.run_until_complete(main())
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/app/sf_ingestion/main.py", line 38, in main
    await produce_case_feed_to_kafka_topic(case_feed_event=case_feed_event)
  File "/usr/local/lib/python3.11/site-packages/fastkafka/_components/producer_decorator.py", line 227, in _produce_async
    await produce_single(producer, topic, encoder_fn, wrapped_val)
  File "/usr/local/lib/python3.11/site-packages/fastkafka/_components/producer_decorator.py", line 107, in produce_single
    fut = await producer.send(
                ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'send'
```

**Reply by kumaranvpl** [reply]
Hello @Jrqz,


How are you running FastKafka application? Ideally you should use [fastkafka run](https://fastkafka.airt.ai/docs/cli/fastkafka/#fastkafka-run) command to run it

For your code, the command would be

`fastkafka run --kafka-broker local --num-workers 1`

**Reply by Jrqz** [reply]
Ah yes, sorry. My config is a bit misleading because the broker is remote in an Azure cloud.

Also while I think of it (although this is a different issue) there seems to be a bug in the FastKafka Tester functionality, in that it fails if the topic ID contains any hyphens (which are allowed by the Kafka standards).

For example, the output, for a topic name of "chatoptin-sf-connector" is as follows (and this error vanishes when the hyphens are replaced by underscores) :

```Traceback (most recent call last):
  File "/usr/local/lib/python3.11/unittest/async_case.py", line 90, in _callTestMethod
    if self._callMaybeAsync(method) is not None:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/unittest/async_case.py", line 112, in _callMaybeAsync
    return self._asyncioRunner.run(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/unittest/mock.py", line 1392, in patched
    return await func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/tests/test_ingestion.py", line 84, in test_main_date_range
    async with Tester(kafka_service.kafka_app):
  File "/usr/local/lib/python3.11/site-packages/fastkafka/_application/tester.py", line 118, in __aenter__
    return await self._ctx.__aenter__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/fastkafka/_application/tester.py", line 104, in _create_ctx
    await self._start_tester()
  File "/usr/local/lib/python3.11/site-packages/fastkafka/_application/tester.py", line 83, in _start_tester
    self.create_mocks()
  File "/usr/local/lib/python3.11/site-packages/fastkafka/_application/app.py", line 1031, in create_mocks
    self.AppMocks = namedtuple(  # type: ignore
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/collections/__init__.py", line 393, in namedtuple
    raise ValueError('Type names and field names must be valid '
ValueError: Type names and field names must be valid identifiers: 'mirror_139850039076816_on_chatoptin-sf-connector_5729812687901956768'```

**Reply by kumaranvpl** [reply]
Hello @Jrqz,

Internally we are using namedtuple in FastKafka Tester. Looks like this error is arising from namedtuple module. We will look into this issue. 

Thanks for letting us know.

**Reply by Jrqz** [reply]
Many thanks Kumaran. It would be good if you (or someone) could also when convenient look into why that producer_store list in producer_decorator.py has its second element as None, when elsewhere the code attempts to use this value as a producer object. For reference, this is the config passed to the FastKafka() class:

```    def get_default_config(self):
        kafka_brokers: dict = {
            "local": {
                "description": "local development kafka broker",
                "url": config.KAFKA_BROKER_URL,
                "port": int(config.KAFKA_BROKER_PORT),
                "protocol": "kafka-secure",
                "security": {"type": "scramSha256"},
            },
        }
        return {
            "title": "SF Ingestion",
            "security_protocol": config.KAFKA_SECURITY_PROTOCOL,
            "sasl_mechanism": config.KAFKA_SASL_MECHANISM,
            "sasl_plain_username": config.KAFKA_BROKER_USERNAME,
            "sasl_plain_password": config.KAFKA_BROKER_PASSWORD,
            "ssl_context": create_ssl_context(),
            "kafka_brokers": kafka_brokers,
        }
```

I tried adding a "production" dict below the "local" dict in the kafka_broker list. But that made no difference

On the assumption that the four element `producer_store` list in producer_decorator.py was maybe (?) supposed to represent the slots for up to four Kafka brokers, I also tried slightly amending the producer_decorator.py code to use the first (and in my example only non-None) element of the producer_store list as the producer object. But that didn't work either. It was rather a silly assumption though, more a stab in the dark, because why would the code always choose only the second element in the list?!

(To repeat what I mentioned earlier, that name "local" for the Kafka broker is misleading because the URL is actually a remote broker!)

**Reply by sternakt** [reply, marked as answer]
Hi!
The failing of tester because of the hyphen in the topic name has been fixed in https://github.com/airtai/fastkafka/issues/472
You can expect it to be working in the next release :)

About the producers_store: the second element in the producers_list is a reference to the concrete producer, this is set to None when you create your application by design, when FastKafka app is started, it will call a populate_producers method which will then replace this None value with the Producer object.
